cumulus/pallets/parachain-system/src/parachain_inherent.rs

```diff
@@ -0,0 +1,527 @@
+			);
+		}
+	}
+}
+
+impl<Message: InboundMessage> Default for AbridgedInboundMessagesCollection<Message> {
+	fn default() -> Self {
+		Self { full_messages: vec![], hashed_messages: vec![] }
+	}
+}
+
+impl InboundMessage for InboundDownwardMessage<RelayChainBlockNumber> {
+	type CompressedMessage = HashedMessage;
+
+	fn data(&self) -> &[u8] {
+		&self.msg
+	}
+
+	fn sent_at(&self) -> RelayChainBlockNumber {
+		self.sent_at
+	}
+
+	fn to_compressed(&self) -> Self::CompressedMessage {
+		self.into()
+	}
+}
+
+pub type InboundDownwardMessages =
+	InboundMessagesCollection<InboundDownwardMessage<RelayChainBlockNumber>>;
+
+pub type AbridgedInboundDownwardMessages =
+	AbridgedInboundMessagesCollection<InboundDownwardMessage<RelayChainBlockNumber>>;
+
+impl AbridgedInboundDownwardMessages {
+	/// Returns an iterator over the messages that maps them to `BoundedSlices`.
+	pub fn bounded_msgs_iter<MaxMessageLen: Get<u32>>(
+		&self,
+	) -> impl Iterator<Item = BoundedSlice<u8, MaxMessageLen>> {
+		self.full_messages
+			.iter()
+			// Note: we are not using `.defensive()` here since that prints the whole value to
+			// console. In case that the message is too long, this clogs up the log quite badly.
+			.filter_map(|m| match BoundedSlice::try_from(&m.msg[..]) {
+				Ok(bounded) => Some(bounded),
+				Err(_) => {
+					defensive!("Inbound Downward message was too long; dropping");
+					None
+				},
+			})
+	}
+}
+
+impl InboundMessage for (ParaId, InboundHrmpMessage) {
+	type CompressedMessage = (ParaId, HashedMessage);
+
+	fn data(&self) -> &[u8] {
+		&self.1.data
+	}
+
+	fn sent_at(&self) -> RelayChainBlockNumber {
+		self.1.sent_at
+	}
+
+	fn to_compressed(&self) -> Self::CompressedMessage {
+		let (sender, message) = self;
+		(*sender, message.into())
+	}
+}
+
+pub type InboundHrmpMessages = InboundMessagesCollection<(ParaId, InboundHrmpMessage)>;
+
+impl InboundHrmpMessages {
+	// Prepare horizontal messages for a more convenient processing:
+	//
+	// Instead of a mapping from a para to a list of inbound HRMP messages, we will have a
+	// list of tuples `(sender, message)` first ordered by `sent_at` (the relay chain block
+	// number in which the message hit the relay-chain) and second ordered by para id
+	// ascending.
+	pub fn from_map(messages_map: BTreeMap<ParaId, Vec<InboundHrmpMessage>>) -> Self {
+		let mut messages = messages_map
+			.into_iter()
+			.flat_map(|(sender, channel_contents)| {
+				channel_contents.into_iter().map(move |message| (sender, message))
+			})
+			.collect::<Vec<_>>();
+		messages.sort_by(|(sender_a, msg_a), (sender_b, msg_b)| {
+			// first sort by sent-at and then by the para id
+			(msg_a.sent_at, sender_a).cmp(&(msg_b.sent_at, sender_b))
+		});
+
+		Self { messages }
+	}
+}
+
+pub type AbridgedInboundHrmpMessages =
+	AbridgedInboundMessagesCollection<(ParaId, InboundHrmpMessage)>;
+
+impl AbridgedInboundHrmpMessages {
+	/// Returns a list of all the unique senders.
+	pub fn get_senders(&self) -> Vec<ParaId> {
```

Comment:
Maybe returning a set here also make the intention clear and simplifies the code a bit.